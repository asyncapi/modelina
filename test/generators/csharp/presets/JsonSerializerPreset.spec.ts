import {
  CSharpGenerator,
  CSHARP_JSON_SERIALIZER_PRESET
} from '../../../../src/generators';
const doc = {
  $id: 'Test',
  type: 'object',
  additionalProperties: true,
  required: ['string prop'],
  properties: {
    'string prop': { type: 'string' },
    'const string prop': { type: 'string', const: 'abc' },
    numberProp: { type: 'number' },
    enumProp: {
      $id: 'EnumTest',
      enum: ['Some enum String', true, { test: 'test' }, 2]
    },
    objectProp: {
      type: 'object',
      $id: 'NestedTest',
      properties: { stringProp: { type: 'string' } }
    }
  },
  patternProperties: {
    '^S(.?)test': {
      type: 'string'
    }
  }
};
describe('JSON serializer preset', () => {
  test('should render serialize and deserialize converters', async () => {
    const generator = new CSharpGenerator({
      presets: [CSHARP_JSON_SERIALIZER_PRESET]
    });

    const inputModel = await generator.generate(doc);
    expect(inputModel).toHaveLength(3);
    expect(inputModel[0].result).toMatchSnapshot();
    expect(inputModel[1].result).toMatchSnapshot();
    expect(inputModel[2].result).toMatchSnapshot();
  });
});
