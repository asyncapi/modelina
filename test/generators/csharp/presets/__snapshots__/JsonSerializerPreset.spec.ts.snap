// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON serializer preset should render serialize and deserialize converters 1`] = `
"[JsonConverter(typeof(TestConverter))]
public partial class Test
{
  private string stringProp;
  private double? numberProp;
  private EnumTest? enumProp;
  private NestedTest? objectProp;
  private Dictionary<string, dynamic>? additionalProperties;

  public string StringProp 
  {
    get { return stringProp; }
    set { this.stringProp = value; }
  }

  public double? NumberProp 
  {
    get { return numberProp; }
    set { this.numberProp = value; }
  }

  public EnumTest? EnumProp 
  {
    get { return enumProp; }
    set { this.enumProp = value; }
  }

  public NestedTest? ObjectProp 
  {
    get { return objectProp; }
    set { this.objectProp = value; }
  }

  public Dictionary<string, dynamic>? AdditionalProperties 
  {
    get { return additionalProperties; }
    set { this.additionalProperties = value; }
  }
}

internal class TestConverter : JsonConverter<Test>
{
  public override bool CanConvert(System.Type objectType)
  {
    // this converter can be applied to any type
    return true;
  }
  public override Test Read(ref Utf8JsonReader reader, System.Type typeToConvert, JsonSerializerOptions options)
  {
    if (reader.TokenType != JsonTokenType.StartObject)
    {
      throw new JsonException();
    }

    var instance = new Test();
  
    while (reader.Read())
    {
      if (reader.TokenType == JsonTokenType.EndObject)
      {
        return instance;
      }

      // Get the key.
      if (reader.TokenType != JsonTokenType.PropertyName)
      {
        throw new JsonException();
      }

      string propertyName = reader.GetString();
      if (propertyName == \\"string prop\\")
        {
          var value = JsonSerializer.Deserialize<string>(ref reader, options);
          instance.StringProp = value;
          continue;
        }
      if (propertyName == \\"numberProp\\")
        {
          var value = JsonSerializer.Deserialize<double?>(ref reader, options);
          instance.NumberProp = value;
          continue;
        }
      if (propertyName == \\"enumProp\\")
        {
          var value = EnumTestExtensions.ToEnumTest(JsonSerializer.Deserialize<dynamic>(ref reader, options));
          instance.EnumProp = value;
          continue;
        }
      if (propertyName == \\"objectProp\\")
        {
          var value = JsonSerializer.Deserialize<NestedTest?>(ref reader, options);
          instance.ObjectProp = value;
          continue;
        }
      if(instance.AdditionalProperties == null) { instance.AdditionalProperties = new Dictionary<string, dynamic>(); }
            var deserializedValue = JsonSerializer.Deserialize<Dictionary<string, dynamic>?>(ref reader, options);
            instance.AdditionalProperties.Add(propertyName, deserializedValue);
            continue;
    }
  
    throw new JsonException();
  }
  public override void Write(Utf8JsonWriter writer, Test value, JsonSerializerOptions options)
  {
    if (value == null)
    {
      JsonSerializer.Serialize(writer, null, options);
      return;
    }
    var properties = value.GetType().GetProperties().Where(prop => prop.Name != \\"AdditionalProperties\\");
  
    writer.WriteStartObject();

    if(value.StringProp != null) {
      // write property name and let the serializer serialize the value itself
      writer.WritePropertyName(\\"string prop\\");
      JsonSerializer.Serialize(writer, value.StringProp, options);
    }
    if(value.NumberProp != null) {
      // write property name and let the serializer serialize the value itself
      writer.WritePropertyName(\\"numberProp\\");
      JsonSerializer.Serialize(writer, value.NumberProp, options);
    }
    if(value.EnumProp != null) {
      // write property name and let the serializer serialize the value itself
      writer.WritePropertyName(\\"enumProp\\");
      JsonSerializer.Serialize(writer, value.EnumProp?.GetValue(), options);
    }
    if(value.ObjectProp != null) {
      // write property name and let the serializer serialize the value itself
      writer.WritePropertyName(\\"objectProp\\");
      JsonSerializer.Serialize(writer, value.ObjectProp, options);
    }
    // Unwrap dictionary properties
    if (value.AdditionalProperties != null) {
      foreach (var unwrappedProperty in value.AdditionalProperties)
      {
        // Ignore any unwrapped properties which might already be part of the core properties
        if (properties.Any(prop => prop.Name == unwrappedProperty.Key))
        {
            continue;
        }
        // Write property name and let the serializer serialize the value itself
        writer.WritePropertyName(unwrappedProperty.Key);
        JsonSerializer.Serialize(writer, unwrappedProperty.Value, options);
      }
    }

    writer.WriteEndObject();
  }

}
"
`;

exports[`JSON serializer preset should render serialize and deserialize converters 2`] = `
"public enum EnumTest
{
  SOME_SPACE_ENUM_SPACE_STRING,
  RESERVED_TRUE,
  CURLYLEFT_QUOTATION_TEST_QUOTATION_COLON_QUOTATION_TEST_QUOTATION_CURLYRIGHT,
  NUMBER_2
}

public static class EnumTestExtensions
{
  public static dynamic? GetValue(this EnumTest enumValue)
  {
    switch (enumValue)
    {
      case EnumTest.SOME_SPACE_ENUM_SPACE_STRING: return \\"Some enum String\\";
      case EnumTest.RESERVED_TRUE: return true;
      case EnumTest.CURLYLEFT_QUOTATION_TEST_QUOTATION_COLON_QUOTATION_TEST_QUOTATION_CURLYRIGHT: return \\"{\\\\\\"test\\\\\\":\\\\\\"test\\\\\\"}\\";
      case EnumTest.NUMBER_2: return 2;
    }
    return null;
  }

  public static EnumTest? ToEnumTest(dynamic? value)
  {
    switch (value)
    {
      case \\"Some enum String\\": return EnumTest.SOME_SPACE_ENUM_SPACE_STRING;
      case true: return EnumTest.RESERVED_TRUE;
      case \\"{\\\\\\"test\\\\\\":\\\\\\"test\\\\\\"}\\": return EnumTest.CURLYLEFT_QUOTATION_TEST_QUOTATION_COLON_QUOTATION_TEST_QUOTATION_CURLYRIGHT;
      case 2: return EnumTest.NUMBER_2;
    }
    return null;
  }
}
"
`;

exports[`JSON serializer preset should render serialize and deserialize converters 3`] = `
"[JsonConverter(typeof(NestedTestConverter))]
public partial class NestedTest
{
  private string? stringProp;
  private Dictionary<string, dynamic>? additionalProperties;

  public string? StringProp 
  {
    get { return stringProp; }
    set { this.stringProp = value; }
  }

  public Dictionary<string, dynamic>? AdditionalProperties 
  {
    get { return additionalProperties; }
    set { this.additionalProperties = value; }
  }
}

internal class NestedTestConverter : JsonConverter<NestedTest>
{
  public override bool CanConvert(System.Type objectType)
  {
    // this converter can be applied to any type
    return true;
  }
  public override NestedTest Read(ref Utf8JsonReader reader, System.Type typeToConvert, JsonSerializerOptions options)
  {
    if (reader.TokenType != JsonTokenType.StartObject)
    {
      throw new JsonException();
    }

    var instance = new NestedTest();
  
    while (reader.Read())
    {
      if (reader.TokenType == JsonTokenType.EndObject)
      {
        return instance;
      }

      // Get the key.
      if (reader.TokenType != JsonTokenType.PropertyName)
      {
        throw new JsonException();
      }

      string propertyName = reader.GetString();
      if (propertyName == \\"stringProp\\")
        {
          var value = JsonSerializer.Deserialize<string?>(ref reader, options);
          instance.StringProp = value;
          continue;
        }
      if(instance.AdditionalProperties == null) { instance.AdditionalProperties = new Dictionary<string, dynamic>(); }
            var deserializedValue = JsonSerializer.Deserialize<Dictionary<string, dynamic>?>(ref reader, options);
            instance.AdditionalProperties.Add(propertyName, deserializedValue);
            continue;
    }
  
    throw new JsonException();
  }
  public override void Write(Utf8JsonWriter writer, NestedTest value, JsonSerializerOptions options)
  {
    if (value == null)
    {
      JsonSerializer.Serialize(writer, null, options);
      return;
    }
    var properties = value.GetType().GetProperties().Where(prop => prop.Name != \\"AdditionalProperties\\");
  
    writer.WriteStartObject();

    if(value.StringProp != null) {
      // write property name and let the serializer serialize the value itself
      writer.WritePropertyName(\\"stringProp\\");
      JsonSerializer.Serialize(writer, value.StringProp, options);
    }
    // Unwrap dictionary properties
    if (value.AdditionalProperties != null) {
      foreach (var unwrappedProperty in value.AdditionalProperties)
      {
        // Ignore any unwrapped properties which might already be part of the core properties
        if (properties.Any(prop => prop.Name == unwrappedProperty.Key))
        {
            continue;
        }
        // Write property name and let the serializer serialize the value itself
        writer.WritePropertyName(unwrappedProperty.Key);
        JsonSerializer.Serialize(writer, unwrappedProperty.Value, options);
      }
    }

    writer.WriteEndObject();
  }

}
"
`;
